---
title: "A short report on the **testReads** file"
author: "Khristian Kotov"
date: "28.02.2015"
output: pdf_document
---
# Prerequisites

Input file [testReads.fastq](http://dl.dropbox.com/u/68829208/testReads.fastq.gz)

1. This file contains NGS reads for some human DNA regions.
2. The fastq file format is described at http://en.wikipedia.org/wiki/FASTQ_format, the
Phred quality score is in sanger format not illumina.
3. The report should be in PDF format and figures rather than descriptive words is preferred. Please also make clear description on your figures.
4. The report could be anything that you find interesting on the file. For example, it could be patterns you found on the data, com

# Exploratory analysis of headers

Each record header consists of five fields:
*unique id*, *rank*, *x*, *y*, and *length*. An example looks as follows:

@HKVCQHG01APQJG rank=0000016 x=176.0 y=10.0 length=108

The basic distributions of these variables do not seem to tell us any particularly interesting story:

```{r, echo=F, message=FALSE, fig.height=4}

if( !file.exists("testReads.fastq") ){
   print("Downloading original data")
   download.file("http://dl.dropbox.com/u/68829208/testReads.fastq.gz", destfile="testReads.fastq.gz")
   ungzip("testReads.fastq.gz")
}

if( !file.exists("qwe.csv") ){
   system("grep '^@' testReads.fastq | sed -e 's|=| |g' -e 's|rank 0*|rank |g' -n -e 's|^@HKVCQHG01||p' | awk 'BEGIN {print \"ID,rank,x,y,len\"} {print $1\",\"$3\",\"$5\",\"$7\",\"$9}' > qwe.csv")
}

library(ggplot2)
qwe <- read.csv("qwe.csv",head=T,sep=',')

p1 <- ggplot(data=qwe,aes(x=rank)) + geom_histogram() + scale_x_continuous(breaks=seq(0,240000,60000))
p2 <- ggplot(data=qwe,aes(x=x)) + geom_histogram()
p3 <- ggplot(data=qwe,aes(x=y)) + geom_histogram()
p4 <- ggplot(data=qwe,aes(x=len)) + geom_histogram(breaks=seq(0,1000,1))

library(gridExtra)
grid.arrange(p1, p4, p2, p3, ncol=2)
```

Neither do correlations tell us an interesting story. So let us omit rather non-informative scatter plots here.
<!-- (except for a weak positive *rank* <-> *length*) -->


# Analysis of the sequences

Analysis design:

- pick a pattern length from 1 to 32 symbols
- scan sequences and identify patters that repeat more often than others
- count all unique patters

Starting from the code that I already developed for the task #1, I implemented [a simple command line tool](https://github.com/koskot77/sophia/blob/master/analysis.cc) that runs a density check for patterns of a specified length. In the trivial case of *length=1* we have four patterns after the four bases:

```{r, echo=F, message=FALSE}

if( !file.exists("sophia/") ){
   system("git clone https://github.com/koskot77/sophia.git")
   system("cd sophia && g++ -Wall -o test analysis.cc && for i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 ; do time ./test $i ; mv output.csv output_$i.csv ; done")
}
an <- read.csv(file='sophia/output_1.csv',head=T,sep=',')
print(an)
```


For second and third trivial cases of *length=2* and *3* the 16 and 64 patterns occur at following rates:

```{r, echo=F, message=FALSE, fig.height=3}
an2 <- read.csv(file='sophia/output_2.csv',header=T,sep=',')
p2  <- ggplot(aes(x=label,y=count),data=an2) + geom_histogram(stat="identity") +
       theme(axis.text.x = element_text(angle=90)) +
       labs(x="pattern",y="count",title="Pattern length = 2")
an3 <- read.csv(file='sophia/output_3.csv',header=T,sep=',')
p3  <- ggplot(aes(x=label,y=count),data=an3) + geom_histogram(stat="identity") +
       theme(axis.text.x = element_text(angle=90)) +
       labs(x="pattern",y="count",title="Pattern length = 3") +
       theme(axis.text.x = element_text(size = 3))
grid.arrange(p2, p3, ncol=2)
```

Number of found unique combinations of the 4 bases grows as 4^*length*^ until *length~10* where it slows down (patterns start look less random while still growing exponentially in numbers):

```{r, echo=F, message=FALSE, fig.height=2.5}
#combinations <- data.frame( width=integer(), found=integer() )
#for(w in 1:32){
#  print(w)
#  an<-read.csv(file=paste('sophia/output_',w,'.csv',sep=""),header=T,sep=',')
#  combinations <- rbind( combinations, data.frame( width=w, found=log2(dim(an)[[1]]) ) )
#}
#print(combinations)
combinations <- data.frame( width=seq(1,32,1), found=c(
2, 4, 6, 8, 10, 12, 13.99, 15.89, 17.36, 18.24, 18.71, 18.97,
19.13, 19.25, 19.36, 19.46, 19.55, 19.63, 19.72, 19.79, 19.87, 19.94,
20.01, 20.07, 20.14, 20.20, 20.26, 20.32, 20.38, 20.43, 20.48, 20.50) )

ggplot(data=combinations,aes(x=width,y=found)) + geom_point() + labs(x="pattern length (symbols)",y="log2(# of unique patterns)",title="Data variability")
```

Let us check how often patterns of specific lengths repeat in the data. From now on for the presentation purposes I use unique *pattern ID* numbers in place of the lengthy symbolic sequences. You can associate the *pattern ID* to a sequence using my more interactive [**shiny application**](https://koskot77.shinyapps.io/sophia/).

```{r, echo=F, message=FALSE, fig.height=3}
an20 <- read.csv(file='sophia/output_20.csv',header=T,sep=',')
p20  <- ggplot(data=an20,aes(x=ID,y=count)) + geom_point(size=1) +
        labs(x="pattern ID",y="count",title="Pattern length = 20")

an10 <- read.csv(file='sophia/output_10.csv',header=T,sep=',')
p10  <- ggplot(data=an10,aes(x=ID,y=count)) + geom_point(size=1) +
        labs(x="pattern ID",y="count",title="Pattern length = 10")

grid.arrange(p10, p20, ncol=2)
```

Interesting features on these plots are horizontal and vertical clusters of dots. The explanation is simple: 
each of the two 20-symbols long patterns from the highest points in the right plot gives 10 different 10-symbols long patterns  that look like the highest horizontal line in the left plot. In other words, longer pattern generates a horizontal line in a narrower view. Similarly, vertical lines correspond to shorter patterns extended by one or two random symbols.

The most frequent patterns between 20 and 32 symbols long are given below (first four are aligned to the common sequence):

length |            sequence              | frequency |
-------|----------------------------------|-----------|
20     | _________AAGACTCGGCAGCATCTCCA___ |    47K    |
21     | ________GAAGACTCGGCAGCATCTCCA___ |    24K    |
26     | TCAGACACGAAGACTCGGCAGCATCT______ |    12K    |
30     | TCAGACACGAAGACTCGGCAGCATCTCCAT__ |    3.8K   |
32     | CTTCCATTGACCACATCTCCTCTGACTTCAAA |    2.2K   |


# Phred quality codes

Distribution of the Phred quality (shifted by 33 units so as to be adjusted to 0) for the most frequent patterns of *length=30* (left plot) looks somewhat better than *length=32* (right plot), but quality is high in both cases:

```{r, echo=F, message=FALSE, fig.height=3}
if( !file.exists("sophia/pattern32.csv") ){
   system("cd sophia && time ./test 32 c CTTCCATTGACCACATCTCCTCTGACTTCAAA ; mv pattern.csv pattern32.csv")
}
if( !file.exists("sophia/pattern30.csv") ){
   system("cd sophia && time ./test 30 c TCAGACACGAAGACTCGGCAGCATCTCCAT ; mv pattern.csv pattern30.csv")
}
q32 <- read.csv(file='sophia/pattern32.csv',header=T,sep=',')
p32 <- ggplot(data=q32,aes(x=qual)) + geom_histogram( breaks=seq(25,40,0.5) ) +
       labs(x=expression(paste("Average quality (",Sigma,q[i]," / length)")),y="# of patterns",title="CTTCCATTGACCACATCTCCTCTGACTTCAAA") +
       theme(plot.title = element_text(size = 8))

q30 <- read.csv(file='sophia/pattern30.csv',header=T,sep=',')
p30 <- ggplot(data=q30,aes(x=qual)) + geom_histogram( breaks=seq(25,40,0.5) ) +
       labs(x=expression(paste("Average quality (",Sigma,q[i]," / length)")),y="# of patterns",title="TCAGACACGAAGACTCGGCAGCATCTCCAT") +
       theme(plot.title = element_text(size = 8))

grid.arrange(p30, p32, ncol=2)
```

Although, there exist patterns grouping at qualities between 20 and 30:

```{r, echo=F, message=FALSE, fig.height=3}

p10 <- ggplot(data=an10,aes(x=meanQual,y=count)) + geom_point(size=1) +
       labs(x=expression(paste("Average quality (",Sigma,q[i]," / length)")),y="# of patterns",title="Pattern length = 10")

p20 <- ggplot(data=an20,aes(x=meanQual,y=count)) + geom_point(size=1) +
       labs(x=expression(paste("Average quality (",Sigma,q[i]," / length)")),y="# of patterns",title="Pattern length = 20")

grid.arrange(p10, p20, ncol=2)
```

(As in the previous plots, the four highest points in the right plot generate four clusters showing up in the left plot. Similarly, clusters in the right plot are generated by even longer patterns.)

# Summary

- Total 84205 records were analyzed.

- 432 records have base labeled 'N' with the corresponding adjusted quality equal to 0. Patterns that included such terms were dismissed from the study.

- Several most common patterns of considerable *length* were identified

- Quality for such patterns looks satisfactory